### PR TITLE
MCR-2173 add consistent error handling to REST API v2

### PR DESCRIPTION
**Pull request by yagee-de**
[Link to jira](https://mycore.atlassian.net/browse/MCR-2173).

### CODE REVIEW

File: mycore-restapi/src/main/java/org/mycore/restapi/v2/MCRErrorResponse.java [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+                    case UNSUPPORTED_MEDIA_TYPE:
+                        e = new NotSupportedException(getMessage(), response, getCause());
+                        break;
+                    default:
+                        e = new ClientErrorException(getMessage(), response, getCause());
+                }
+                break;
+            case SERVER_ERROR:
+                //Response.Status.OK is to trigger "default" case
+                switch (s != null ? s : Response.Status.OK) {
```
Comment (adlerre) [inline mycore-restapi/src/main/java/org/mycore/restapi/v2/MCRErrorResponse.java:119]: Codacy Issue found: [A switch statement does not contain a break](https://app.codacy.com/manual/MyCoRe/mycore/pullRequest?prid=5532334)

**Comment by adlerre** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Complexity increasing per file
==============================
- mycore-restapi/src/main/java/org/mycore/restapi/v2/MCRErrorResponse.java  15
- mycore-restapi/src/main/java/org/mycore/restapi/v2/MCRRestAuthorizationFilter.java  7
- mycore-restapi/src/main/java/org/mycore/restapi/v1/MCRRestAuthorizationFilter.java  7
- mycore-restapi/src/main/java/org/mycore/restapi/v2/MCRExceptionMapper.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/MyCoRe/mycore/pullRequest?prid=5532334&bid=18135174)

File: mycore-restapi/src/main/java/org/mycore/restapi/MCRRestFeature.java [hunk 2]
```diff
@@ -86,7 +86,11 @@ protected void registerTransactionFilter(FeatureContext context) {
 
     @Override
     protected void registerAccessFilter(FeatureContext context, Class<?> resourceClass, Method resourceMethod) {
-        context.register(MCRRestAuthorizationFilter.class);
+        if (resourceClass.getPackageName().contains(".v1")) {
+            context.register(org.mycore.restapi.v1.MCRRestAuthorizationFilter.class);
+        } else {
+            context.register(org.mycore.restapi.v2.MCRRestAuthorizationFilter.class);
```
Comment (adlerre) [inline mycore-restapi/src/main/java/org/mycore/restapi/MCRRestFeature.java:92]: Codacy Issue found: [Unnecessary use of fully qualified name 'org.mycore.restapi.v2.MCRRestAuthorizationFilter' due to existing same package import 'org.mycore.restapi.*'](https://app.codacy.com/manual/MyCoRe/mycore/pullRequest?prid=5532334)

File: mycore-restapi/src/main/java/org/mycore/restapi/MCRRestFeature.java [hunk 3]
```diff
@@ -86,7 +86,11 @@ protected void registerTransactionFilter(FeatureContext context) {
 
     @Override
     protected void registerAccessFilter(FeatureContext context, Class<?> resourceClass, Method resourceMethod) {
-        context.register(MCRRestAuthorizationFilter.class);
+        if (resourceClass.getPackageName().contains(".v1")) {
+            context.register(org.mycore.restapi.v1.MCRRestAuthorizationFilter.class);
```
Comment (adlerre) [inline mycore-restapi/src/main/java/org/mycore/restapi/MCRRestFeature.java:90]: Codacy Issue found: [Unnecessary use of fully qualified name 'org.mycore.restapi.v1.MCRRestAuthorizationFilter' due to existing same package import 'org.mycore.restapi.*'](https://app.codacy.com/manual/MyCoRe/mycore/pullRequest?prid=5532334)

File: mycore-restapi/src/main/java/org/mycore/restapi/v2/MCRErrorResponse.java [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+    public void setDetail(String detail) {
+        this.detail = detail;
+    }
+
+    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = MCRRestUtils.JSON_DATE_FORMAT)
+    @XmlAttribute
+    public Date getTimestamp() {
+        return Date.from(timestamp);
+    }
+
```
Comment (rsteph-de) [inline mycore-restapi/src/main/java/org/mycore/restapi/v2/MCRErrorResponse.java:190]: provide an adapter:
```
public class MCRInstantXMLAdapter extends XmlAdapter<String, Instant>{

    @Override
    public Instant unmarshal(String v) throws Exception {
        return Instant.parse(v);
    }

    @Override
    public String marshal(Instant v) throws Exception {
       return v.toString();
    }
}
```

and annote like this:
```
  @XmlAttribute
    @XmlJavaTypeAdapter(MCRInstantXMLAdapter.class)
     public Instant getTimestamp() { ...}
```
works for JSON as well

File: mycore-restapi/src/main/java/org/mycore/restapi/v2/MCRRestObjects.java [hunk 5]
```diff
@@ -274,7 +272,10 @@ public Response getObjectVersions(@Parameter(example = "mir_mods_00004711") @Pat
         throws IOException {
         long modified = MCRXMLMetadataManager.instance().getLastModified(id);
         if (modified < 0) {
-            throw new NotFoundException("MCRObject " + id + " not found");
+            throw MCRErrorResponse.fromStatus(Response.Status.NOT_FOUND.getStatusCode())
+                .withErrorCode("MCROBJECT_NOT_FOUND")
```
Comment (rsteph-de) [inline mycore-restapi/src/main/java/org/mycore/restapi/v2/MCRRestObjects.java]: Would it make sense to collect all error codes as static variables in a central class?

**Comment by adlerre** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 4
           

Complexity increasing per file
==============================
- mycore-restapi/src/main/java/org/mycore/restapi/v2/MCRExceptionMapper.java  2
- mycore-restapi/src/main/java/org/mycore/restapi/v2/MCRRestErrorCodes.java  1
- mycore-restapi/src/main/java/org/mycore/restapi/v1/MCRRestAuthorizationFilter.java  7
- mycore-restapi/src/main/java/org/mycore/restapi/v2/MCRErrorResponse.java  15
- mycore-restapi/src/main/java/org/mycore/restapi/v2/MCRRestAuthorizationFilter.java  7
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/MyCoRe/mycore/pullRequest?prid=5532334&bid=18135174)

File: mycore-restapi/src/main/java/org/mycore/restapi/MCRRestFeature.java (hunk 2 above)
Comment (adlerre) [inline mycore-restapi/src/main/java/org/mycore/restapi/MCRRestFeature.java:92]: Codacy Issue found: [Unnecessary use of fully qualified name 'org.mycore.restapi.v2.MCRRestAuthorizationFilter' due to existing same package import 'org.mycore.restapi.*'](https://app.codacy.com/manual/MyCoRe/mycore/pullRequest?prid=5532334)

File: mycore-restapi/src/main/java/org/mycore/restapi/MCRRestFeature.java (hunk 3 above)
Comment (adlerre) [inline mycore-restapi/src/main/java/org/mycore/restapi/MCRRestFeature.java:90]: Codacy Issue found: [Unnecessary use of fully qualified name 'org.mycore.restapi.v1.MCRRestAuthorizationFilter' due to existing same package import 'org.mycore.restapi.*'](https://app.codacy.com/manual/MyCoRe/mycore/pullRequest?prid=5532334)

File: mycore-restapi/src/main/java/org/mycore/restapi/v2/MCRErrorResponse.java (hunk 1 above)
Comment (adlerre) [inline mycore-restapi/src/main/java/org/mycore/restapi/v2/MCRErrorResponse.java:119]: Codacy Issue found: [A switch statement does not contain a break](https://app.codacy.com/manual/MyCoRe/mycore/pullRequest?prid=5532334)

File: mycore-restapi/src/main/java/org/mycore/restapi/v2/MCRRestErrorCodes.java [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *  
+ * You should have received a copy of the GNU General Public License
+ * along with MyCoRe.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package org.mycore.restapi.v2;
+
+final class MCRRestErrorCodes {
```
Comment (adlerre) [inline mycore-restapi/src/main/java/org/mycore/restapi/v2/MCRRestErrorCodes.java:21]: Codacy Issue found: [The utility class name 'MCRRestErrorCodes' doesn't match '[A-Z][a-zA-Z0-9]+(Utils?|Helper|Constants)'](https://app.codacy.com/manual/MyCoRe/mycore/pullRequest?prid=5532334)

**Comment by adlerre** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 4
           

Complexity increasing per file
==============================
- mycore-restapi/src/main/java/org/mycore/restapi/v2/MCRExceptionMapper.java  2
- mycore-restapi/src/main/java/org/mycore/restapi/v2/MCRRestErrorCodes.java  1
- mycore-restapi/src/main/java/org/mycore/restapi/v2/MCRRestAuthorizationFilter.java  7
- mycore-restapi/src/main/java/org/mycore/restapi/v2/MCRErrorResponse.java  15
- mycore-restapi/src/main/java/org/mycore/restapi/v1/MCRRestAuthorizationFilter.java  7
- mycore-restapi/src/main/java/org/mycore/restapi/converter/MCRInstantXMLAdapter.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/MyCoRe/mycore/pullRequest?prid=5532334&bid=18135174)

File: mycore-restapi/src/main/java/org/mycore/restapi/MCRRestFeature.java (hunk 2 above)
Comment (adlerre) [inline mycore-restapi/src/main/java/org/mycore/restapi/MCRRestFeature.java:92]: Codacy Issue found: [Unnecessary use of fully qualified name 'org.mycore.restapi.v2.MCRRestAuthorizationFilter' due to existing same package import 'org.mycore.restapi.*'](https://app.codacy.com/manual/MyCoRe/mycore/pullRequest?prid=5532334)

File: mycore-restapi/src/main/java/org/mycore/restapi/MCRRestFeature.java (hunk 3 above)
Comment (adlerre) [inline mycore-restapi/src/main/java/org/mycore/restapi/MCRRestFeature.java:90]: Codacy Issue found: [Unnecessary use of fully qualified name 'org.mycore.restapi.v1.MCRRestAuthorizationFilter' due to existing same package import 'org.mycore.restapi.*'](https://app.codacy.com/manual/MyCoRe/mycore/pullRequest?prid=5532334)

File: mycore-restapi/src/main/java/org/mycore/restapi/v2/MCRErrorResponse.java [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+                    case UNSUPPORTED_MEDIA_TYPE:
+                        e = new NotSupportedException(getMessage(), response, getCause());
+                        break;
+                    default:
+                        e = new ClientErrorException(getMessage(), response, getCause());
+                }
+                break;
+            case SERVER_ERROR:
+                //Response.Status.OK is to trigger "default" case
+                switch (s != null ? s : Response.Status.OK) {
```
Comment (adlerre) [inline mycore-restapi/src/main/java/org/mycore/restapi/v2/MCRErrorResponse.java:120]: Codacy Issue found: [A switch statement does not contain a break](https://app.codacy.com/manual/MyCoRe/mycore/pullRequest?prid=5532334)

File: mycore-restapi/src/main/java/org/mycore/restapi/v2/MCRRestErrorCodes.java (hunk 6 above)
Comment (adlerre) [inline mycore-restapi/src/main/java/org/mycore/restapi/v2/MCRRestErrorCodes.java:21]: Codacy Issue found: [The utility class name 'MCRRestErrorCodes' doesn't match '[A-Z][a-zA-Z0-9]+(Utils?|Helper|Constants)'](https://app.codacy.com/manual/MyCoRe/mycore/pullRequest?prid=5532334)

**Comment by yagee-de** [comment]
> why not simply Instant.toString()
> https://docs.oracle.com/javase/8/docs/api/java/time/Instant.html#toString--

Will provide this step in the following commit. Thanks for the hint!

**Comment by adlerre** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 4
           

Complexity increasing per file
==============================
- mycore-restapi/src/main/java/org/mycore/restapi/v2/MCRRestErrorCodes.java  1
- mycore-restapi/src/main/java/org/mycore/restapi/v1/MCRRestAuthorizationFilter.java  7
- mycore-restapi/src/main/java/org/mycore/restapi/v2/MCRExceptionMapper.java  2
- mycore-restapi/src/main/java/org/mycore/restapi/v2/MCRErrorResponse.java  15
- mycore-restapi/src/main/java/org/mycore/restapi/converter/MCRInstantXMLAdapter.java  1
- mycore-restapi/src/main/java/org/mycore/restapi/v2/MCRRestAuthorizationFilter.java  7
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/MyCoRe/mycore/pullRequest?prid=5532334&bid=18135174)

File: mycore-restapi/src/main/java/org/mycore/restapi/v2/MCRErrorResponse.java (hunk 7 above)
Comment (adlerre) [inline mycore-restapi/src/main/java/org/mycore/restapi/v2/MCRErrorResponse.java:120]: Codacy Issue found: [A switch statement does not contain a break](https://app.codacy.com/manual/MyCoRe/mycore/pullRequest?prid=5532334)

File: mycore-restapi/src/main/java/org/mycore/restapi/v2/MCRRestErrorCodes.java (hunk 6 above)
Comment (adlerre) [inline mycore-restapi/src/main/java/org/mycore/restapi/v2/MCRRestErrorCodes.java:21]: Codacy Issue found: [The utility class name 'MCRRestErrorCodes' doesn't match '[A-Z][a-zA-Z0-9]+(Utils?|Helper|Constants)'](https://app.codacy.com/manual/MyCoRe/mycore/pullRequest?prid=5532334)

File: mycore-restapi/src/main/java/org/mycore/restapi/MCRRestFeature.java (hunk 2 above)
Comment (adlerre) [inline mycore-restapi/src/main/java/org/mycore/restapi/MCRRestFeature.java:92]: Codacy Issue found: [Unnecessary use of fully qualified name 'org.mycore.restapi.v2.MCRRestAuthorizationFilter' due to existing same package import 'org.mycore.restapi.*'](https://app.codacy.com/manual/MyCoRe/mycore/pullRequest?prid=5532334)

File: mycore-restapi/src/main/java/org/mycore/restapi/MCRRestFeature.java (hunk 3 above)
Comment (adlerre) [inline mycore-restapi/src/main/java/org/mycore/restapi/MCRRestFeature.java:90]: Codacy Issue found: [Unnecessary use of fully qualified name 'org.mycore.restapi.v1.MCRRestAuthorizationFilter' due to existing same package import 'org.mycore.restapi.*'](https://app.codacy.com/manual/MyCoRe/mycore/pullRequest?prid=5532334)

**Comment by adlerre** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Complexity increasing per file
==============================
- mycore-restapi/src/main/java/org/mycore/restapi/v2/MCRErrorResponse.java  15
- mycore-restapi/src/main/java/org/mycore/restapi/v2/MCRErrorCodeConstants.java  1
- mycore-restapi/src/main/java/org/mycore/restapi/v2/MCRExceptionMapper.java  2
- mycore-restapi/src/main/java/org/mycore/restapi/v2/MCRRestAuthorizationFilter.java  7
- mycore-restapi/src/main/java/org/mycore/restapi/converter/MCRInstantXMLAdapter.java  1
- mycore-restapi/src/main/java/org/mycore/restapi/v1/MCRRestAuthorizationFilter.java  7
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/MyCoRe/mycore/pullRequest?prid=5532334&bid=18135174)

File: mycore-restapi/src/main/java/org/mycore/restapi/v2/MCRErrorResponse.java (hunk 7 above)
Comment (adlerre) [inline mycore-restapi/src/main/java/org/mycore/restapi/v2/MCRErrorResponse.java:120]: Codacy Issue found: [A switch statement does not contain a break](https://app.codacy.com/manual/MyCoRe/mycore/pullRequest?prid=5532334)

File: mycore-restapi/src/main/java/org/mycore/restapi/MCRRestFeature.java (hunk 2 above)
Comment (adlerre) [inline mycore-restapi/src/main/java/org/mycore/restapi/MCRRestFeature.java:92]: Codacy Issue found: [Unnecessary use of fully qualified name 'org.mycore.restapi.v2.MCRRestAuthorizationFilter' due to existing same package import 'org.mycore.restapi.*'](https://app.codacy.com/manual/MyCoRe/mycore/pullRequest?prid=5532334)

File: mycore-restapi/src/main/java/org/mycore/restapi/MCRRestFeature.java (hunk 3 above)
Comment (adlerre) [inline mycore-restapi/src/main/java/org/mycore/restapi/MCRRestFeature.java:90]: Codacy Issue found: [Unnecessary use of fully qualified name 'org.mycore.restapi.v1.MCRRestAuthorizationFilter' due to existing same package import 'org.mycore.restapi.*'](https://app.codacy.com/manual/MyCoRe/mycore/pullRequest?prid=5532334)

**Comment by MyCoRe-Travis** [comment]
Something gone wrong on d3cbefa410c2569fa1084738880a9b3044567a2f! 
 Please Check https://github.com/MyCoRe-Travis/test_artifacts/tree/5580-pull_request-969

**Pull request merged**